### PR TITLE
#1864 If the coordinates have a centroid that is far from zero, the molecule placed far offscreen

### PR DESCRIPTION
**Pull request by porcelain11**
refactor Render class, add types

### CODE REVIEW

File: packages/ketcher-core/src/application/editor/operations/CanvasLoad.ts [hunk 1]
```diff
@@ -32,7 +32,10 @@ export class CanvasLoad extends BaseOperation {
   execute(restruct: ReStruct) {
     const oldStruct = restruct.molecule
     restruct.clearVisels() // TODO: What is it?
```
Comment (KonstantinEpam23) [inline packages/ketcher-core/src/application/editor/operations/CanvasLoad.ts:34]: This clears up visuals elements of all Re* entities

File: packages/ketcher-core/src/application/render/restruct/restruct.ts [hunk 2]
```diff
@@ -575,7 +575,7 @@ class ReStruct {
     })
   }
 
-  setSelection(selection) {
+  setSelection(selection?) {
     const redraw = arguments.length === 0 // render.update only
```
Comment (KonstantinEpam23) [inline packages/ketcher-core/src/application/render/restruct/restruct.ts:579]: if (selection) ?

Comment (KonstantinEpam23) [reply packages/ketcher-core/src/application/render/restruct/restruct.ts:579]: no idea why we do this thing with `arguments`

Comment (KonstantinEpam23) [inline packages/ketcher-core/src/application/render/restruct/restruct.ts:579]: would be good to get rid of counting args at some point

File: packages/ketcher-core/src/application/editor/operations/CanvasLoad.ts (hunk 1 above)
Comment (porcelain11) [reply packages/ketcher-core/src/application/editor/operations/CanvasLoad.ts:34]: This comment is not mine - but, actually, for me is also it is not clear, why do we need to clear them

File: packages/ketcher-core/src/application/render/raphaelRender.ts [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+import { Box2Abs, Struct, Vec2 } from 'domain/entities'
+
+import Raphael from './raphael-ext'
+import { ReStruct } from './restruct'
+import { Scale } from 'domain/helpers'
+import defaultOptions from './options'
+import draw from './draw'
+import { RaphaelPaper } from 'raphael'
+// import { RenderOptions } from './render.types'
```
Comment (epam-nikolai-iakushchenko) [inline packages/ketcher-core/src/application/render/raphaelRender.ts:25]: leftover?

File: packages/ketcher-core/src/application/render/raphaelRender.ts [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+  x1: number,
+  y1: number
+) {
+  // eslint-disable-line max-params
+  let ex = x0 < 0 ? -x0 : 0
+  let ey = y0 < 0 ? -y0 : 0
+
+  if (scaledSz.x < x1) ex += x1 - scaledSz.x
+  if (scaledSz.y < y1) ey += y1 - scaledSz.y
+  return new Vec2(ex, ey)
```
Comment (epam-nikolai-iakushchenko) [inline packages/ketcher-core/src/application/render/raphaelRender.ts:40]:
```suggestion
  if (scaledSz.y < y1) ey += y1 - scaledSz.y
  
  return new Vec2(ex, ey)
```

File: packages/ketcher-core/src/application/render/raphaelRender.ts [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+  public clientArea: HTMLElement
+  private userOpts: any
+  public paper: RaphaelPaper
+  private sz: Vec2
+  public ctab: ReStruct
+  public options: any
+  private renderWidth: number
+  private renderHeight: number
+  private useOldZoom: boolean
+  private oldCb: Box2Abs | null
```
Comment (epam-nikolai-iakushchenko) [inline packages/ketcher-core/src/application/render/raphaelRender.ts:53]: Not pushing, but if you know what sz, ctab, oldCb mean, would be great to name them something meaningful 😄

File: packages/ketcher-core/src/application/render/raphaelRender.ts [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+
+  obj2view(v: Vec2, isRelative: boolean) {
+    let p = Scale.obj2scaled(v, this.options)
+    p = isRelative
+      ? p
+      : p
+          .add(this.options.offset)
+          .sub(this.getScrollPos().scaled(1 / this.options.zoom))
+    if (!this.useOldZoom) p = p.scaled(this.options.zoom)
+    return p
```
Comment (epam-nikolai-iakushchenko) [inline packages/ketcher-core/src/application/render/raphaelRender.ts:117]:
```suggestion
    if (!this.useOldZoom) p = p.scaled(this.options.zoom)
    
    return p
```

File: packages/ketcher-core/src/application/render/raphaelRender.ts [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+  }
+
+  page2obj(event: MouseEvent) {
+    const clientArea = this.clientArea
+    const { top: offsetTop, left: offsetLeft } =
+      clientArea.getBoundingClientRect()
+
+    const pp = new Vec2(event.clientX - offsetLeft, event.clientY - offsetTop)
+    return this.view2obj(pp)
```
Comment (epam-nikolai-iakushchenko) [inline packages/ketcher-core/src/application/render/raphaelRender.ts:126]:
```suggestion
    const pp = new Vec2(event.clientX - offsetLeft, event.clientY - offsetTop)
   
     return this.view2obj(pp)
```

File: packages/ketcher-core/src/application/render/raphaelRender.ts [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+        )
+        const cb = Box2Abs.union(viewBox, extendedBox)
+        if (!this.oldCb) {
+          this.oldCb = new Box2Abs()
+        }
+
+        const sz = cb.sz().floor()
+        const delta = this.oldCb.p0.sub(cb.p0).ceil()
+        // TODO: do we need this.oldBb? Seem it's not in use
+        // this.oldBb = bb
```
Comment (epam-nikolai-iakushchenko) [inline packages/ketcher-core/src/application/render/raphaelRender.ts:239]: looks like commented-out code 😄

File: packages/ketcher-core/src/application/render/restruct/restruct.ts [hunk 9]
```diff
@@ -605,9 +604,11 @@ class ReStruct {
             )
             item.selected = sGroupAtoms.length > 0 && sGroupAtoms[0].selected
           }
-          const selected = redraw
-            ? item.selected
-            : selection && selection[map] && selection[map].indexOf(id) > -1
+
+          const selected =
+            selection && selection[map]
```
Comment (epam-nikolai-iakushchenko) [inline packages/ketcher-core/src/application/render/restruct/restruct.ts:609]:
```suggestion
            selection?.[map]
```

**Review by epam-nikolai-iakushchenko — APPROVED** [review]

File: packages/ketcher-core/src/application/render/raphaelRender.ts (hunk 8 above)
Comment (porcelain11) [reply packages/ketcher-core/src/application/render/raphaelRender.ts:239]: Yes - a bit scary to remove anything completely - probably, it was needed for something

File: packages/ketcher-core/src/application/render/restruct/restruct.ts (hunk 9 above)
Comment (porcelain11) [reply packages/ketcher-core/src/application/render/restruct/restruct.ts:609]: Fixed

File: packages/ketcher-core/src/application/render/raphaelRender.ts (hunk 7 above)
Comment (porcelain11) [reply packages/ketcher-core/src/application/render/raphaelRender.ts:126]: Fixed

File: packages/ketcher-core/src/application/render/raphaelRender.ts (hunk 6 above)
Comment (porcelain11) [reply packages/ketcher-core/src/application/render/raphaelRender.ts:117]: Fixed

File: packages/ketcher-core/src/application/render/raphaelRender.ts (hunk 4 above)
Comment (porcelain11) [reply packages/ketcher-core/src/application/render/raphaelRender.ts:40]: Fixed

**Pull request merged**